File: api_nestjs_distributed_order_platform/src/infrastructure/payment/stub-payment.gateway.ts
import { Injectable } from "@nestjs/common";
import type { PaymentGatewayInterface } from "src/application/port/payment-gateway.repository";
import type { GatewayResult } from "src/shared/interfaces/gateway.interface";
import { randomUUID } from "crypto";

@Injectable()
export class StubPaymentGateway implements PaymentGatewayInterface {
    private readonly charges = new Map<string, GatewayResult>();

    async createCharge(input: {
        paymentId: string;
        orderId: string;
        amount: number;
        idempotencyKey: string;
    }): Promise<GatewayResult> {
        const existing = this.charges.get(input.idempotencyKey);
        if (existing) {
            return existing;
        }

        const cents = Math.round((input.amount % 1) * 100);

        if (cents === 13) {
            throw new Error('ETIMEDOUT: gateway connection timeout');
        }

        if (cents === 99) {
            const failed: GatewayResult = {
                success: false,
                kind: 'BUSINESS',
                status: 'failed',
                errorCode: 'card_declined',
                errorMessage: 'Card was declined',
                raw: { reason: 'card_declined' },
            };
            this.charges.set(input.idempotencyKey, failed);
            return failed;
        }

        const paid: GatewayResult = {
            success: true,
            status: 'paid',
            externalId: `stub_${randomUUID()}`,
            raw: {
                paymentId: input.paymentId,
                orderId: input.orderId,
                amount: input.amount,
            },
        };
        this.charges.set(input.idempotencyKey, paid);
        return paid;
    }

    async retrieveCharge(input: {
        idempotencyKey: string;
        externalId?: string;
    }): Promise<GatewayResult> {
        const existing = this.charges.get(input.idempotencyKey);
        if (existing) {
            return existing;
        }

        return {
            success: false,
            kind: 'RETRIABLE',
            status: 'pending',
            errorCode: 'CHARGE_NOT_FOUND',
            errorMessage: 'Charge not found yet',
        };
    }
}

File: api_java_distributed_order_platform/src/main/java/com/project/order/infrastructure/payment/StubPaymentGateway.java
package com.project.order.infrastructure.payment;

import com.project.order.application.exception.GatewayException;
import com.project.order.application.port.PaymentGatewayRepository;
import com.project.order.shared.interfaces.GatewayResult;
import org.springframework.stereotype.Component;

import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Espelha o StubPaymentGateway do NestJS: sem chamar nenhuma API de
 * verdade, usa os centavos do valor pra simular os três cenários (timeout,
 * cartão recusado, sucesso), guardando o resultado por idempotencyKey pra
 * que um retry devolva sempre a mesma resposta.
 */
@Component
public class StubPaymentGateway implements PaymentGatewayRepository {

    private final Map<String, GatewayResult> charges = new ConcurrentHashMap<>();

    @Override
    public GatewayResult createCharge(String paymentId, String orderId, double amount, String idempotencyKey) {
        GatewayResult existing = charges.get(idempotencyKey);
        if (existing != null) {
            return existing;
        }

        int cents = (int) Math.round((amount % 1) * 100);

        if (cents == 13) {
            throw new GatewayException("ETIMEDOUT: gateway connection timeout");
        }

        if (cents == 99) {
            GatewayResult failed = new GatewayResult(false, null, "failed", "card_declined", "Card was declined", "BUSINESS",
                    Map.of("reason", "card_declined"));
            charges.put(idempotencyKey, failed);
            return failed;
        }

        GatewayResult paid = new GatewayResult(true, "stub_" + UUID.randomUUID(), "paid", null, null, null,
                Map.of("paymentId", paymentId, "orderId", orderId, "amount", amount));
        charges.put(idempotencyKey, paid);
        return paid;
    }

    @Override
    public GatewayResult retrieveCharge(String idempotencyKey, String externalId) {
        GatewayResult existing = charges.get(idempotencyKey);
        if (existing != null) {
            return existing;
        }
        return new GatewayResult(false, null, "pending", "CHARGE_NOT_FOUND", "Charge not found yet", "RETRIABLE", null);
    }
}
